replica/database.cc

```diff
@@ -1283,13 +1284,24 @@ database::existing_index_names(const sstring& ks_name, const sstring& cf_to_excl
 
 future<std::tuple<lw_shared_ptr<query::result>, cache_temperature>>
 database::query(schema_ptr s, const query::read_command& cmd, query::result_options opts, const dht::partition_range_vector& ranges,
-                tracing::trace_state_ptr trace_state, db::timeout_clock::time_point timeout) {
+                tracing::trace_state_ptr trace_state, db::timeout_clock::time_point timeout, db::allow_per_partition_rate_limit allow_limit) {
     const auto reversed = cmd.slice.is_reversed();
     if (reversed) {
         s = s->make_reversed();
     }
 
     column_family& cf = find_column_family(cmd.cf_id);
+
+    if (allow_limit && ranges.front().is_singular()) {
+        if (auto table_limit = s->per_partition_rate_limit_options().get_max_reads_per_second()) {
+            auto& read_label = cf.get_rate_limiter_label_for_reads();
+            auto token = dht::token::to_int64(ranges.front().start()->value().token());
```

Comment:
I changed the logic so that all singular ranges are accounted.

However, I think there is another problem with my current implementation. Now, replicas reject reads with some probability and for each singular range the decision is made separately. A read will be rejected if the decision is "reject" for _any_ of the partitions, so the combined probability of rejection is higher than any separate partition.

I'm not sure how should we behave in this case. Compute the probabilities of rejection, take maximum and then toss a coin based on that? Or maybe not rate limit at all, to be consistent with scans (for which I don't think there is a sensible way to apply per-partition rate limit)?